sources/zarr/large_image_source_zarr/__init__.py

```diff
@@ -438,6 +492,208 @@ def getTile(self, x, y, z, pilImageAllowed=False, numpyAllowed=False, **kwargs):
+        to accommodate it.  Note that x and y can be negative.  If so, the
+        output image (and internal memory access of the image) will act as if
+        the 0, 0 point is the most negative position.  Cropping is applied
+        after this offset.
+
+        :param tile: a numpy array, PIL Image, or a binary string
+            with an image.  The numpy array can have 2 or 3 dimensions.
+        :param x: location in destination for upper-left corner.
+        :param y: location in destination for upper-left corner.
+        :param mask: a 2-d numpy array (or 3-d if the last dimension is 1).
+            If specified, areas where the mask is false will not be altered.
+        :param axes: a string or list of strings specifying the names of axes
+            in the same order as the tile dimensions
+        :param kwargs: start locations for any additional axes
+        """
+        # TODO: improve band bookkeeping
+
+        self._checkEditable()
+        placement = {
+            'x': x,
+            'y': y,
+            **kwargs,
+        }
+        if not isinstance(tile, np.ndarray) or axes is None:
+            axes = 'yxs'
+            tile, mode = _imageToNumpy(tile)
+        elif not isinstance(axes, str) and not isinstance(axes, list):
+            err = 'Invalid type for axes. Must be str or list[str].'
+            raise ValueError(err)
+        axes = [x.lower() for x in axes]
+        if axes[-1] != 's':
+            axes.append('s')
+        if 'x' not in axes or 'y' not in axes:
+            err = 'Invalid value for axes. Must contain "y" and "x".'
+            raise ValueError(err)
+        self._axes = {k: i for i, k in enumerate(axes)}
+
+        while len(tile.shape) < len(axes):
+            tile = np.expand_dims(tile, axis=0)
+
+        new_dims = {
+            a: max(
+                self._dims.get(a, 0),
+                placement.get(a, 0) + tile.shape[i],
+            )
+            for a, i in self._axes.items()
+        }
+        self._dims = new_dims
+        self._dtype = tile.dtype
+        self._bandCount = new_dims.get(axes[-1])  # last axis is assumed to be bands
+        self.sizeX = new_dims.get('x')
+        self.sizeY = new_dims.get('y')
+        self._framecount = np.prod([
+            length
+            for axis, length in new_dims.items()
+            if axis in axes[:-3]
+        ])
+        self._levels = None
+        self.levels = int(max(1, math.ceil(math.log(max(
+            self.sizeX / self.tileWidth, self.sizeY / self.tileHeight)) / math.log(2)) + 1))
+
+        placement_slices = tuple([
+            slice(placement.get(a, 0), placement.get(a, 0) + tile.shape[i], 1)
+            for i, a in enumerate(axes)
+        ])
+        if mask is not None and len(mask.shape) + 1 == len(tile.shape):
+            mask = mask[:, :, np.newaxis]
+
+        current_arrays = dict(self._zarr.arrays())
+        with self._addLock:
```

Comment:
We need to make sure any access to `self.<attr>` that could change is inside of the `_addLock`.  Moving it above `self._axes = ...` would work, but maybe some other lines could be moved outside of the lock.

Reply:
I moved all the changes to the internal attributes to the end of `addTile`, within the `addLock` context, here: [6706f69](https://github.com/girder/large_image/pull/1446/commits/6706f6988764ffcbc02886e9ebd5e4d5333c7496). Is this what you mean?

Reply:
Mostly.  The `self._axes = ` (above) needs to be within the lock context, too.

Reply:
Changed in [32c9c58](https://github.com/girder/large_image/pull/1446/commits/32c9c5819c60276749f3beebf0c5d36ef848eeab)